core/src/main/java/brooklyn/location/access/PortForwardManager.java

```diff
@@ -46,6 +47,42 @@
 @Beta
 public interface PortForwardManager extends Location {
 
+    @Beta
+    class AssociationMetadata {
+        private final String publicIpId;
+        private final HostAndPort publicEndpoint;
+        private final Location location;
+        private final int privatePort;
+
+        public AssociationMetadata(String publicIpId, HostAndPort publicEndpoint, Location location, int privatePort) {
+            this.publicIpId = publicIpId;
+            this.publicEndpoint = publicEndpoint;
+            this.location = location;
+            this.privatePort = privatePort;
+        }
+
+        public String getPublicIpId() {
+            return publicIpId;
+        }
+
+        public HostAndPort getPublicEndpoint() {
+            return publicEndpoint;
+        }
+
+        public Location getLocation() {
+            return location;
+        }
+
+        public int getPrivatePort() {
+            return privatePort;
+        }
+    }
+
+    @Beta
+    interface AssociationListener {
+        void onEvent(AssociationMetadata metadata);
```

Comment:
Do you think we should call this on removal of a port mapping? I think we probably should.

The reason is that it impacts the choice of method name, so impacts backwards compatibility. How about changing this to `onAssociationCreated(AssociationMetadata)` and `onAssociationDeleted(AssociationMetadata)`?

For the impl of the latter, see `PortForwardManagerImpl.forgetPortMapping(...)` and `forgetPortMappings(...)` - i.e. everywhere that it removes from the `PortForwardManagerImpl.mappings` map.